Build fails (with recent EUMM?)

I see build failures on smoker machines where Crypt-MatrixSSL3-v3.7.6 formerly built and tested OK. Excerpt of the log:
```
cc -c  -I. -Iinc/matrixssl-3-7-2b-open -Iinc/matrixssl-3-7-2b-open/crypto -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -Os -Wp,-w -fomit-frame-pointer -fPIC -I./ -I../ -O2   -DVERSION=\"v3.7.6\" -DXS_VERSION=\"v3.7.6\" -fPIC "-I/opt/perl-5.18.2t/lib/5.18.2/x86_64-linux-thread-multi/CORE"  -DUSE_TLS_RSA_WITH_AES_128_CBC_SHA -DUSE_TLS_RSA_WITH_AES_256_CBC_SHA -DUSE_TLS_RSA_WITH_AES_128_CBC_SHA256 -DUSE_TLS_RSA_WITH_AES_256_CBC_SHA256 -DUSE_TLS_RSA_WITH_AES_128_GCM_SHA256 -DUSE_TLS_RSA_WITH_AES_256_GCM_SHA384 -DUSE_TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA -DUSE_TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA -DUSE_TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384 -DUSE_TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256 -DUSE_TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -DUSE_TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 -DUSE_TLS_DHE_PSK_WITH_AES_256_CBC_SHA -DUSE_TLS_DHE_PSK_WITH_AES_128_CBC_SHA -DUSE_TLS_DHE_RSA_WITH_AES_256_CBC_SHA -DUSE_TLS_DHE_RSA_WITH_AES_128_CBC_SHA -DUSE_TLS_DHE_RSA_WITH_AES_128_CBC_SHA256 -DUSE_TLS_DHE_RSA_WITH_AES_256_CBC_SHA256 -DUSE_AES -DUSE_AES_GCM -DUSE_3DES -DUSE_DES -DUSE_ARC4 -DUSE_RC2 -DUSE_SHA1 -DUSE_SHA256 -DUSE_SHA384 -DUSE_SHA512 -DUSE_MD5 -DUSE_HMAC -DUSE_ECC -DUSE_DH -DUSE_RSA -DUSE_X509 -DUSE_BASE64_DECODE -DUSE_CERT_PARSE -DUSE_FULL_CERT_PARSE -DUSE_PRIVATE_KEY_PARSING -DMATRIX_USE_FILE_SYSTEM -DUSE_CLIENT_SIDE_SSL -DUSE_SERVER_SIDE_SSL -DUSE_TLS -DUSE_TLS_1_1 -DUSE_TLS_1_2 -DDISABLE_SSLV3 -DUSE_STATELESS_SESSION_TICKETS -DENABLE_SECURE_REHANDSHAKES -DENABLE_TLS_FALLBACK_SCSV -DREQUIRE_SECURE_REHANDSHAKES -DENABLE_CERTIFICATE_STATUS_REQUEST -DENABLE_SIGNED_CERTIFICATE_TIMESTAMP -DUSE_ALPN -DENABLE_FALSE_START -DUSE_BEAST_WORKAROUND -DUSE_CLIENT_AUTH -DSERVER_CAN_SEND_EMPTY_CERT_REQUEST -DSERVER_WILL_ACCEPT_EMPTY_CLIENT_CERT_MSG -DUSE_PKCS5 -DUSE_PKCS8 -DUSE_PKCS12 -DPS_PUBKEY_OPTIMIZE_FOR_FASTER_SPEED -DPS_AES_IMPROVE_PERF_INCREASE_CODESIZE -DPS_3DES_IMPROVE_PERF_INCREASE_CODESIZE -DPS_MD5_IMPROVE_PERF_INCREASE_CODESIZE -DPS_SHA1_IMPROVE_PERF_INCREASE_CODESIZE -DLINUX -DPOSIX -DPSTM_64BIT  MatrixSSL3.c 
rm -f blib/arch/auto/Crypt/MatrixSSL3/MatrixSSL3.so
cc  -shared -O2 -L/usr/local/lib -fstack-protector -o blib/arch/auto/Crypt/MatrixSSL3/MatrixSSL3.so  inc/matrixssl-3-7-2b-open/core/corelib.o inc/matrixssl-3-7-2b-open/core/memset_s.o inc/matrixssl-3-7-2b-open/core/POSIX/osdep.o inc/matrixssl-3-7-2b-open/crypto/digest/hmac.o inc/matrixssl-3-7-2b-open/crypto/digest/md2.o inc/matrixssl-3-7-2b-open/crypto/digest/md4.o inc/matrixssl-3-7-2b-open/crypto/digest/md5.o inc/matrixssl-3-7-2b-open/crypto/digest/sha1.o inc/matrixssl-3-7-2b-open/crypto/digest/sha224.o inc/matrixssl-3-7-2b-open/crypto/digest/sha256.o inc/matrixssl-3-7-2b-open/crypto/digest/sha384.o inc/matrixssl-3-7-2b-open/crypto/digest/sha512.o inc/matrixssl-3-7-2b-open/crypto/hardware/aesni.o inc/matrixssl-3-7-2b-open/crypto/keyformat/asn1.o inc/matrixssl-3-7-2b-open/crypto/keyformat/base64.o inc/matrixssl-3-7-2b-open/crypto/keyformat/x509.o inc/matrixssl-3-7-2b-open/crypto/math/pstm.o inc/matrixssl-3-7-2b-open/crypto/math/pstm_montgomery_reduce.o inc/matrixssl-3-7-2b-open/crypto/math/pstm_mul_comba.o inc/matrixssl-3-7-2b-open/crypto/math/pstm_sqr_comba.o inc/matrixssl-3-7-2b-open/crypto/prng/prng.o inc/matrixssl-3-7-2b-open/crypto/prng/yarrow.o inc/matrixssl-3-7-2b-open/crypto/pubkey/dh.o inc/matrixssl-3-7-2b-open/crypto/pubkey/ecc.o inc/matrixssl-3-7-2b-open/crypto/pubkey/pkcs.o inc/matrixssl-3-7-2b-open/crypto/pubkey/pubkey.o inc/matrixssl-3-7-2b-open/crypto/pubkey/rsa.o inc/matrixssl-3-7-2b-open/crypto/symmetric/aes.o inc/matrixssl-3-7-2b-open/crypto/symmetric/aesGCM.o inc/matrixssl-3-7-2b-open/crypto/symmetric/arc4.o inc/matrixssl-3-7-2b-open/crypto/symmetric/des3.o inc/matrixssl-3-7-2b-open/crypto/symmetric/idea.o inc/matrixssl-3-7-2b-open/crypto/symmetric/rc2.o inc/matrixssl-3-7-2b-open/crypto/symmetric/seed.o inc/matrixssl-3-7-2b-open/matrixssl/cipherSuite.o inc/matrixssl-3-7-2b-open/matrixssl/dtls.o inc/matrixssl-3-7-2b-open/matrixssl/hsHash.o inc/matrixssl-3-7-2b-open/matrixssl/matrixssl.o inc/matrixssl-3-7-2b-open/matrixssl/matrixsslApi.o inc/matrixssl-3-7-2b-open/matrixssl/prf.o inc/matrixssl-3-7-2b-open/matrixssl/psk.o inc/matrixssl-3-7-2b-open/matrixssl/sslDecode.o inc/matrixssl-3-7-2b-open/matrixssl/sslEncode.o inc/matrixssl-3-7-2b-open/matrixssl/sslv3.o inc/matrixssl-3-7-2b-open/matrixssl/tls.o MatrixSSL3.o  \
	   -lpthread   \
	  
cc: error: inc/matrixssl-3-7-2b-open/core/corelib.o: No such file or directory
cc: error: inc/matrixssl-3-7-2b-open/core/memset_s.o: No such file or directory
cc: error: inc/matrixssl-3-7-2b-open/core/POSIX/osdep.o: No such file or directory
cc: error: inc/matrixssl-3-7-2b-open/crypto/digest/hmac.o: No such file or directory
...
cc: error: inc/matrixssl-3-7-2b-open/matrixssl/sslv3.o: No such file or directory
cc: error: inc/matrixssl-3-7-2b-open/matrixssl/tls.o: No such file or directory
make: *** [blib/arch/auto/Crypt/MatrixSSL3/MatrixSSL3.so] Error 1
```
It seems that this happens with newer ExtUtils::MakeMaker versions (e.g. 7.14) (@bingos: FYI)

--- a/Makefile.PL
+++ b/Makefile.PL
@@ -489,9 +489,9 @@ sub c_o {
     my $inherited = shift->SUPER::c_o(@_);
 
     if ($^O =~ /Win32/ms && $Config{cc} eq 'cl') {  # Microsoft deprecated -o in favor of /Fo
-        $inherited =~ s{\$\*\.c[\s\t]*$}{/Fo\$\*.obj \$\*.c\n}msg;
+        $inherited =~ s{\$\*\.c\s*$}{/Fo\$\*.obj \$\*.c\n}msg;
     } else {
-        $inherited =~ s{\$\*\.c[\s\t]*$}{\$\*.c -o \$\*.o\n}msg;
+        $inherited =~ s{\$\*\.c\s*$}{\$\*.c -o \$\*.o\n}msg;
     }
 
     return $inherited;

--- a/Makefile.PL
+++ b/Makefile.PL
@@ -195,18 +195,21 @@ my @cOptions = (
     # matrixsslConfig.h - Applicable to servers only. The size of the
     # session resumption table for caching session identifiers. Old entries
     # will be overwritten when size is reached
-    #'SSL_SESSION_TABLE_SIZE=32',
+    'SSL_SESSION_TABLE_SIZE=32',
 
     # matrixsslConfig.h - Applicable to servers only. The time in seconds
     # that a session identifier will be valid in the session table. A value of 0
-    # will disable SSL resumption
-    #'SSL_SESSION_ENTRY_LIFE=86400000',
+    # will disable SSL resumption (in milliseconds)
+    'SSL_SESSION_ENTRY_LIFE=86400000',
 
     # matrixsslConfig.h - Enable the stateless ticket resumption mechanism
     # defined by RFC 5077. Clients must then set the
     # ticketResumption member of the session options to 1 to
     # activate for a given session.
     'USE_STATELESS_SESSION_TICKETS',
+
+    # matrixsslConfig.h - the max size of the server key list
+    'SSL_SESSION_TICKET_LIST_LEN=32',
 
     # matrixsslConfig.h - Enable secure rehandshaking as defined in RFC 5746
     'ENABLE_SECURE_REHANDSHAKES',

--- a/Makefile.PL
+++ b/Makefile.PL
@@ -37,7 +37,7 @@ my @cOptions = (
     # - enable USE_DH in the options below (enabled by default)
     # - load DH params for the session keys by using $keys->load_DH_params( $DH_params_file )
     #
-    # If you plan to support HTTP/2 you are require by specification to suuport the following cipher
+    # If you plan to support HTTP/2 you are required by specification to support the following cipher
     #   TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
 
     #******************************************************************************
@@ -360,6 +360,9 @@ if (-d 'inc/patches') {
         for my $file (glob 'inc/patches/*.patch') {
             # replacing \n with \n will change \n to \r\n on Win, without this patch.exe will crash
             system "cd \Q$mxbase\E && perl -pe \"s/\\n/\\n/\" ../../\Q$file\E | patch -p1 -N";
+
+            # some Windows systems don't like the path escape ...
+            # system "cd $mxbase && perl -pe \"s/\\n/\\n/\" ../../$file | patch -p1 -N";
         }
     }
     else {
@@ -486,9 +489,9 @@ sub c_o {
     my $inherited = shift->SUPER::c_o(@_);
 
     if ($^O =~ /Win32/ms && $Config{cc} eq 'cl') {  # Microsoft deprecated -o in favor of /Fo
-        $inherited =~ s{\$\*\.c\n}{/Fo\$\*.obj \$\*.c\n}msg;
+        $inherited =~ s{\$\*\.c[\s\t]*$}{/Fo\$\*.obj \$\*.c\n}msg;
     } else {
-        $inherited =~ s{\$\*\.c\n}{\$\*.c -o \$\*.o\n}msg;
+        $inherited =~ s{\$\*\.c[\s\t]*$}{\$\*.c -o \$\*.o\n}msg;
     }
 
     return $inherited;

--- a/MatrixSSL3.xs
+++ b/MatrixSSL3.xs
@@ -1068,6 +1068,31 @@ int keys_load_ecc(keys, certFile, privFile, privPass, trustedCAcertFiles)
     RETVAL
 
 
+int keys_load_ecc_mem(keys, cert, priv, trustedCA)
+    Crypt_MatrixSSL3_Keys *keys;
+    SV *cert;
+    SV *priv;
+    SV *trustedCA;
+    unsigned char *certBuf = NULL;
+    unsigned char *privBuf = NULL;
+    unsigned char *trustedCABuf = NULL;
+    STRLEN certLen = 0;
+    STRLEN privLen = 0;
+    STRLEN trustedCALen = 0;
+
+    CODE:
+    /* All bufs can contain \0, so SvPV must be used instead of strlen() */
+    certBuf = SvOK(cert) ? (unsigned char *) SvPV(cert, certLen) : NULL;
+    privBuf = SvOK(priv) ? (unsigned char *) SvPV(priv, privLen) : NULL;
+    trustedCABuf= SvOK(trustedCA) ? (unsigned char *) SvPV(trustedCA, trustedCALen) : NULL;
+
+    RETVAL = matrixSslLoadEcKeysMem((sslKeys_t *)keys, certBuf, certLen, privBuf, privLen,
+                                      trustedCABuf, trustedCALen);
+
+    OUTPUT:
+    RETVAL
+
+
 int keys_load_pkcs12(keys, p12File, importPass, macPass, flags)
     Crypt_MatrixSSL3_Keys *keys;
     char *p12File = SvOK(ST(1)) ? SvPV_nolen(ST(1)) : NULL;

--- a/MatrixSSL3.xs
+++ b/MatrixSSL3.xs
@@ -757,6 +757,7 @@ void Close()
     matrixSslClose();
     matrixssl_initialized = 0;
 
+
 int refresh_OCSP_staple(server_index, index, DERfile)
     int server_index = SvOK(ST(0)) ? SvIV(ST(0)) : -1;
     int index = SvOK(ST(1)) ? SvIV(ST(1)): -1;
@@ -862,6 +863,82 @@ int refresh_SCT_buffer(server_index, index, SCT_params)
     warn("Refreshed SCT buffer: Loaded %d SCT files, %p %d", rc, *p_buffer, *p_size);
 #endif
     RETVAL = rc;
+
+    OUTPUT:
+    RETVAL
+
+
+int refresh_ALPN_data(server_index, index, ALPN_data)
+    int server_index = SvOK(ST(0)) ? SvIV(ST(0)) : -1;
+    int index = SvOK(ST(1)) ? SvIV(ST(1)): -1;
+    SV *ALPN_data;
+    AV *aproto = NULL;
+    p_ALPN_data alpn = NULL, *palpn = NULL;
+    SV *tmp_sv = NULL;
+    unsigned char *item = NULL;
+    STRLEN item_len = 0;
+    int i = 0;
+
+    CODE:
+    if (!(SvROK(ALPN_data) && SvTYPE(SvRV(ALPN_data)) == SVt_PVAV))
+        croak("Expected ALPN data to be an array reference");
+
+    if (server_index < 0)
+        croak("Invalid default server/SNI server index %d", server_index);
+
+    if (index < 0) {
+#ifdef MATRIX_DEBUG
+        warn("Refresh default server %d ALPN data", server_index);
+#endif
+        if (server_index >= default_server_index)
+            croak("Out of range default server index specified: %d > %d", server_index, default_server_index - 1);
+
+        palpn = &(default_servers[server_index]->alpn);
+    } else {
+#ifdef MATRIX_DEBUG
+        warn("Refresh ALPN_data for SNI server %d/SNI entry %d", server_index, index);
+#endif
+        if (server_index >= SNI_server_index)
+            croak("Out of range SNI server index spcified: %d > %d", server_index, SNI_server_index - 1);
+
+        if (index >= SNI_servers[server_index]->SNI_entries_number)
+            croak("Out of range SNI entry index spcified for SNI server %d: %d > %d", server_index, index, SNI_servers[server_index]->SNI_entries_number - 1);
+
+        palpn = &(SNI_servers[server_index]->SNI_entries[index]->alpn);
+    }
+
+    /* Check if we should allocate the ALPN data strcture */
+    if (*palpn == NULL) {
+        *palpn = (p_ALPN_data) malloc(SZ_ALPN_DATA);
+        memset(*palpn, 0, SZ_ALPN_DATA);
+    }
+
+    alpn = *palpn;
+    aproto = (AV *) SvRV(ALPN_data);
+
+    /* Free previous allocated protocols (if any) */
+#ifdef MATRIX_DEBUG
+    warn("Freeing %d protocols", alpn->protoCount);
+#endif
+    for (i = 0; i > alpn->protoCount; i++)
+        if (alpn->proto[i]) free(alpn->proto[i]);
+
+    /* Load new protocols */
+    alpn->protoCount = (short) av_len(aproto) + 1;
+    if (alpn->protoCount > MAX_PROTO_EXT) alpn->protoCount = MAX_PROTO_EXT;
+
+    for (i = 0; i < alpn->protoCount; i++) {
+        tmp_sv = *av_fetch(aproto, i, 0);
+        item = (unsigned char *) SvPV(tmp_sv, item_len);
+#ifdef MATRIX_DEBUG
+        warn("Protocol %d: %.*s", i, item_len, item);
+#endif
+        alpn->proto[i] = (unsigned char *) malloc(item_len);
+        memcpy(alpn->proto[i], item, item_len);
+        alpn->protoLen[i] = item_len;
+    }
+
+    RETVAL = alpn->protoCount;
 
     OUTPUT:
     RETVAL

--- a/MatrixSSL3.xs
+++ b/MatrixSSL3.xs
@@ -1189,6 +1189,14 @@ Crypt_MatrixSSL3_Sess *sess_new_server(keys, certValidator)
     OUTPUT:
     RETVAL
 
+SV *get_master_secret(ssl)
+    Crypt_MatrixSSL3_Sess *ssl;
+
+    CODE:
+    RETVAL = newSVpv(matrixSslGetMasterSecret(ssl), SSL_HS_MASTER_SIZE);
+
+    OUTPUT:
+    RETVAL
 
 int sess_init_SNI(ssl, index, ssl_id, sni_data = NULL)
     Crypt_MatrixSSL3_Sess *ssl;

--- a/MatrixSSL3.xs
+++ b/MatrixSSL3.xs
@@ -1054,6 +1054,20 @@ int keys_load_rsa_mem(keys, cert, priv, trustedCA)
     RETVAL
 
 
+int keys_load_ecc(keys, certFile, privFile, privPass, trustedCAcertFiles)
+    Crypt_MatrixSSL3_Keys *keys;
+    char *certFile = SvOK(ST(1)) ? SvPV_nolen(ST(1)) : NULL;
+    char *privFile = SvOK(ST(2)) ? SvPV_nolen(ST(2)) : NULL;
+    char *privPass = SvOK(ST(3)) ? SvPV_nolen(ST(3)) : NULL;
+    char *trustedCAcertFiles = SvOK(ST(4)) ? SvPV_nolen(ST(4)) : NULL;
+
+    CODE:
+    RETVAL = (int) matrixSslLoadEcKeys((sslKeys_t *)keys, certFile, privFile, privPass, trustedCAcertFiles);
+
+    OUTPUT:
+    RETVAL
+
+
 int keys_load_pkcs12(keys, p12File, importPass, macPass, flags)
     Crypt_MatrixSSL3_Keys *keys;
     char *p12File = SvOK(ST(1)) ? SvPV_nolen(ST(1)) : NULL;
@@ -1421,8 +1435,30 @@ int sess_init_SNI(ssl, index, ssl_id, sni_data = NULL)
                     croak("SNI matrixSslLoadRsaKeys failed %d; %s; %s", rc, cert, key);
             } else
                 croak("Bad cert/key specified in SNI entry %d", i);
+        } else if (hv_exists(sd, "ecc_cert", strlen("ecc_cert")) && hv_exists(sd, "ecc_key", strlen("ecc_key"))) {
+            cert_sv = *hv_fetch(sd, "ecc_cert", strlen("ecc_cert"), 0);
+            key_sv = *hv_fetch(sd, "ecc_key", strlen("ecc_key"), 0);
+
+            if (SvOK(cert_sv) && SvOK(key_sv)) {
+                cert = (unsigned char *) SvPV_nolen(cert_sv);
+                key = (unsigned char *) SvPV_nolen(key_sv);
+#ifdef MATRIX_DEBUG
+                warn("  SNI entry %d ecc_cert %s; ecc_key %s", i, cert, key);
+#endif
+                add_obj();
+                rc = matrixSslNewKeys(&(ss->SNI_entries[i]->keys), NULL);
+                if (rc != PS_SUCCESS) {
+                    del_obj();
+                    croak("SNI matrixSslNewKeys failed %d", rc);
+                }
+
+                rc = matrixSslLoadEcKeys(ss->SNI_entries[i]->keys, cert, key, NULL, NULL);
+                if (rc != PS_SUCCESS)
+                    croak("SNI matrixSslLoadEcKeys failed %d; %s; %s", rc, cert, key);
+            } else
+                croak("Bad ECC cert/key specified in SNI entry %d", i);
         } else
-            croak("Missing cert/key specified in SNI entry %d", i);
+            croak("Missing [ECC] cert/key specified in SNI entry %d", i);
 
         if (hv_exists(sd, "DH_param", strlen("DH_param"))) {
             item_sv = *hv_fetch(sd, "DH_param", strlen("DH_param"), 0);

--- a/examples/functions.pl
+++ b/examples/functions.pl
@@ -45,12 +45,15 @@ sub ssl_io {
     # After that we should try to send outgoing data (if any).
     my $err = q{};
 
+    # SSL data buffer
+    my $buf;
+
 RECV:
     # Fill MatrixSSL's read buffer with received data. Repeat until all
     # received data will be moved from our buffer to MatrixSSL.
     while (length $in) {
         # Process (part of) received data:
-        my $rc = $ssl->received_data($in, my $buf);
+        my $rc = $ssl->received_data($in, $buf);
 RC:
         # - "Success. This return code will be returned if $n is 0 and
         #   there is no remaining internal data to process. This could
@@ -188,6 +191,41 @@ SEND:
         #   this return code, a resumed handshake has just completed."
         elsif ($rc == MATRIXSSL_HANDSHAKE_COMPLETE) {
             $handshakeIsComplete = 1;
+
+            # Sometimes clients use the "false start" technique - they send
+            # the first application data packet immediately after sending the
+            # handshake "Finished" message without waiting for the server's own
+            # handshake "Finished" message. This saves a rountrip and speeds the
+            # data transfer in some cases by 33%
+
+            my $rc2 = $ssl->false_start_received_data($buf);
+
+            # all the codes are already explained in this function
+            while (1) {
+                if ($rc2 == MATRIXSSL_APP_DATA) {
+                    $appIn .= $buf if (defined($buf));
+                }
+                elsif ($rc2 == MATRIXSSL_RECEIVED_ALERT) {
+                    my ($level, $descr) = get_ssl_alert($buf);
+                    if ($level == SSL_ALERT_LEVEL_FATAL) {
+                        $ssl->processed_data($buf);
+                        $err = alert($level, $descr);
+                        last;
+                    } else {
+                        warn alert($level, $descr);
+                    }
+                }
+                elsif (($rc2 == MATRIXSSL_SUCCESS) || ($rc2 == MATRIXSSL_REQUEST_SEND) || ($rc2 == MATRIXSSL_REQUEST_RECV)) {
+                    last;
+                }
+                else {
+                    $err = error($rc);
+                    last;
+                }
+
+                $rc2 = $ssl->processed_data($buf);
+            }
+
             next;
         }
         # - "Success. This indicates the message that was sent to the peer

--- a/examples/functions.pl
+++ b/examples/functions.pl
@@ -203,7 +203,7 @@ SEND:
             # all the codes are already explained in this function
             while (1) {
                 if ($rc2 == MATRIXSSL_APP_DATA) {
-                    $appIn .= $buf if (defined($buf));
+                    $appIn .= $buf;
                 }
                 elsif ($rc2 == MATRIXSSL_RECEIVED_ALERT) {
                     my ($level, $descr) = get_ssl_alert($buf);

--- a/inc/patches/core_WIN32_osdep.c.patch
+++ b/inc/patches/core_WIN32_osdep.c.patch
@@ -1,0 +1,18 @@
+--- matrixssl-3-7-2b-open-original/core/WIN32/osdep.c	2015-03-18 20:57:05.000000000 +0000
++++ matrixssl-3-7-2b-open/core/WIN32/osdep.c	2016-03-30 16:04:02.000000000 +0000
+@@ -68,13 +68,13 @@ int32 psGetTime(psTime_t *t, void *userP
+ 	if (t == NULL) {
+ 		QueryPerformanceCounter(&lt);
+ 		diff = lt.QuadPart - hiresStart.QuadPart;
+-		d = (int32)((diff * 1000) / hiresFreq.QuadPart);
++		d = (int32)(diff / hiresFreq.QuadPart);
+ 		return d;
+ 	}
+ 
+ 	QueryPerformanceCounter(t);
+ 	diff = t->QuadPart - hiresStart.QuadPart;
+-	d = (int32)((diff * 1000) / hiresFreq.QuadPart);
++	d = (int32)(diff / hiresFreq.QuadPart);
+ 	return d;
+ }
+ 

--- a/inc/patches/matrixssl_matrixssl.c.patch
+++ b/inc/patches/matrixssl_matrixssl.c.patch
@@ -1,5 +1,5 @@
 --- matrixssl-3-7-2b-open-original/matrixssl/matrixssl.c	2015-03-18 20:57:06.000000000 +0000
-+++ matrixssl-3-7-2b-open/matrixssl/matrixssl.c	2016-03-27 16:07:02.000000000 +0000
++++ matrixssl-3-7-2b-open/matrixssl/matrixssl.c	2016-03-30 16:42:37.000000000 +0000
 @@ -56,7 +56,12 @@ static psMutex_t			g_sessTicketLock;
  #endif
  #endif /* USE_MULTITHREADING */
@@ -92,7 +92,36 @@
  /*
  	The cipher and mac contexts are inline in the ssl structure, so
  	clearing the structure clears those states as well.
-@@ -2107,7 +2142,7 @@ int32 matrixServerSetKeysSNI(ssl_t *ssl,
+@@ -1992,6 +2027,8 @@ int32 matrixResumeSession(ssl_t *ssl)
+ 	unsigned char	*id;
+ 	uint32	i;
+ 
++	psTraceInfo("Trying to resume session\n");
++
+ 	if (!(ssl->flags & SSL_FLAGS_SERVER)) {
+ 		return PS_ARG_FAIL;
+ 	}
+@@ -2015,6 +2052,9 @@ int32 matrixResumeSession(ssl_t *ssl)
+ 	Expiration is done on daily basis (86400 seconds)
+ */
+ 	psGetTime(&accessTime, ssl->userPtr);
++
++	psTraceIntInfo("Session cache still ok for %d ms\n", SSL_SESSION_ENTRY_LIFE - psDiffMsecs(sessionTable[i].startTime, accessTime, ssl->userPtr));
++
+ 	if ((memcmp(sessionTable[i].id, id,
+ 			(uint32)min(ssl->sessionIdLen, SSL_MAX_SESSION_ID_SIZE)) != 0) ||
+ 			(psDiffMsecs(sessionTable[i].startTime,	accessTime, ssl->userPtr) >
+@@ -2040,6 +2080,9 @@ int32 matrixResumeSession(ssl_t *ssl)
+ #else /* PKCS11 TODO */
+ 	psTraceInfo("Session resumption not supported in PKCS#11\n");
+ #endif
++
++	psTraceInfo("Session resumed\n");
++
+ 	return PS_SUCCESS;
+ }
+ 
+@@ -2107,7 +2150,7 @@ int32 matrixServerSetKeysSNI(ssl_t *ssl,
  	if (ssl->sni_cb) {
  		ssl->extFlags.sni = 1; /* extension was actually handled */
  		keys = NULL;
@@ -101,7 +130,7 @@
  		if (keys) {
  			ssl->keys = keys;
  			return 0;
-@@ -2495,7 +2530,7 @@ int32 matrixUnlockSessionTicket(ssl_t *s
+@@ -2495,11 +2538,13 @@ int32 matrixUnlockSessionTicket(ssl_t *s
  	cipherSuite += *enc; enc++;
  
  	/* Force cipher suite */
@@ -110,3 +139,25 @@
  		psTraceInfo("Cipher suite check failure on session ticket\n");
  		return PS_FAILURE;
  	}
+ 
++	psTraceIntInfo("Ticket using cipher %d\n", cipherSuite);
++
+ 	/* Set aside masterSecret */
+ 	memcpy(ssl->sid->masterSecret, enc, SSL_HS_MASTER_SIZE);
+ 	enc += SSL_HS_MASTER_SIZE;
+@@ -2512,11 +2557,15 @@ int32 matrixUnlockSessionTicket(ssl_t *s
+ 
+ 	now = psGetTime(&t, ssl->userPtr);
+ 
++	psTraceIntInfo("Time = %d  ", time); psTraceIntInfo("Now = %d\n", now);
++	psTraceIntInfo("Session ticket still good for %d seconds\n", (SSL_SESSION_ENTRY_LIFE / 1000) - (now - time));
++
+ 	if ((now - time) > (SSL_SESSION_ENTRY_LIFE / 1000)) {
+ 		/* Expired session ticket.  New one will be issued */
+ 		psTraceInfo("Session ticket was expired\n");
+ 		return PS_FAILURE;
+ 	}
++
+ 	ssl->sid->cipherId = cipherSuite;
+ 
+ 	return PS_SUCCESS;

--- a/inc/patches/matrixssl_matrixsslApi.c.patch
+++ b/inc/patches/matrixssl_matrixsslApi.c.patch
@@ -1,6 +1,21 @@
 --- matrixssl-3-7-2b-open-original/matrixssl/matrixsslApi.c	2015-03-18 20:57:06.000000000 +0000
-+++ matrixssl-3-7-2b-open/matrixssl/matrixsslApi.c	2016-03-23 22:32:46.000000000 +0000
-@@ -215,6 +215,98 @@ void matrixSslDeleteSessionId(sslSession
++++ matrixssl-3-7-2b-open/matrixssl/matrixsslApi.c	2016-03-29 22:44:59.000000000 +0000
+@@ -36,6 +36,14 @@
+ 
+ /******************************************************************************/
+ /*
++	Return SSL session master secret
++*/
++unsigned char *matrixSslGetMasterSecret(ssl_t *ssl) {
++	return ssl->sec.masterSecret;
++}
++
++/******************************************************************************/
++/*
+ 	Create a new client SSL session
+ 	This creates internal SSL buffers and cipher structures
+ 	Clients initiate the connection with a 'HelloRequest', and this data
+@@ -215,6 +223,98 @@ void matrixSslDeleteSessionId(sslSession
  #endif /* USE_CLIENT_SIDE_SSL */
  
  #ifdef USE_SERVER_SIDE_SSL
@@ -99,7 +114,7 @@
  /******************************************************************************/
  /*
  	Create a new server SSL session
-@@ -288,18 +380,22 @@ NEW_SVR_ERROR:
+@@ -288,18 +388,22 @@ NEW_SVR_ERROR:
  }
  
  void matrixSslRegisterSNICallback(ssl_t *ssl, void (*sni_cb)(void *ssl,

--- a/inc/patches/matrixssl_matrixsslApi.h.patch
+++ b/inc/patches/matrixssl_matrixsslApi.h.patch
@@ -1,6 +1,14 @@
 --- matrixssl-3-7-2b-open-original/matrixssl/matrixsslApi.h	2015-03-18 20:57:06.000000000 +0000
-+++ matrixssl-3-7-2b-open/matrixssl/matrixsslApi.h	2016-03-23 22:32:46.000000000 +0000
-@@ -182,6 +182,13 @@ PSPUBLIC int32 matrixSslCreateALPNext(ps
++++ matrixssl-3-7-2b-open/matrixssl/matrixsslApi.h	2016-03-29 22:45:30.000000000 +0000
+@@ -136,6 +136,7 @@ PSPUBLIC void	matrixSslDeleteSession(ssl
+ /*
+ 	Advanced feature public APIS
+  */
++PSPUBLIC unsigned char *matrixSslGetMasterSecret(ssl_t *ssl);
+ PSPUBLIC void	matrixSslGetAnonStatus(ssl_t *ssl, int32 *anonArg);
+ PSPUBLIC int32	matrixSslEncodeRehandshake(ssl_t *ssl, sslKeys_t *keys,
+ 					int32 (*certCb)(ssl_t *ssl, psX509Cert_t *cert,int32 alert),
+@@ -182,6 +183,13 @@ PSPUBLIC int32 matrixSslCreateALPNext(ps
  /*
  	Server side APIs
  */
@@ -14,7 +22,7 @@
  PSPUBLIC int32 matrixSslNewServerSession(ssl_t **ssl, const sslKeys_t *keys,
  				int32 (*certCb)(ssl_t *ssl, psX509Cert_t *cert, int32 alert),
  				sslSessOpts_t *options);
-@@ -189,13 +196,13 @@ PSPUBLIC int32 matrixSslSetCipherSuiteEn
+@@ -189,13 +197,13 @@ PSPUBLIC int32 matrixSslSetCipherSuiteEn
  				uint32 status);
  PSPUBLIC void matrixSslRegisterSNICallback(ssl_t *ssl,
  				void (*sni_cb)(void *ssl, char *hostname, int32 hostnameLen,

--- a/inc/patches/matrixssl_matrixsslConfig.h.patch
+++ b/inc/patches/matrixssl_matrixsslConfig.h.patch
@@ -1,63 +1,72 @@
---- matrixssl-3-7-2b-open-original/matrixssl/matrixsslConfig.h	2015-03-18 20:57:06.000000000 +0000
-+++ matrixssl-3-7-2b-open/matrixssl/matrixsslConfig.h	2016-03-30 16:42:19.000000000 +0000
+--- matrixssl-3-7-2b-open-original/matrixssl/matrixsslConfig.h  2015-03-18 20:57:06.000000000 +0000
++++ matrixssl-3-7-2b-open/matrixssl/matrixsslConfig.h   2016-05-10 17:23:49.591000002 +0000
 @@ -61,9 +61,9 @@ extern "C" {
- 	the handshake will determine which is best for the connection.
+        the handshake will determine which is best for the connection.
  */
- 
+
 -#define USE_TLS_RSA_WITH_AES_128_CBC_SHA
 +/* #define USE_TLS_RSA_WITH_AES_128_CBC_SHA */
  #define USE_TLS_RSA_WITH_AES_256_CBC_SHA
 -#define USE_TLS_RSA_WITH_AES_128_CBC_SHA256 /* TLS 1.2 */
-+/* #define USE_TLS_RSA_WITH_AES_128_CBC_SHA256 */ /* TLS 1.2 */ 
++/* #define USE_TLS_RSA_WITH_AES_128_CBC_SHA256 */ /* TLS 1.2 */
  #define USE_TLS_RSA_WITH_AES_256_CBC_SHA256 /* TLS 1.2 */
  /* #define USE_TLS_RSA_WITH_AES_128_GCM_SHA256 */ /* TLS 1.2 */
  /* #define USE_TLS_RSA_WITH_AES_256_GCM_SHA384 */ /* TLS 1.2 */
-@@ -224,14 +224,14 @@ extern "C" {
- 	SSL_SESSION_ENTRY_LIFE is in milliseconds, minimum 0
- 
+@@ -224,14 +224,24 @@ extern "C" {
+        SSL_SESSION_ENTRY_LIFE is in milliseconds, minimum 0
+
  */
--#define SSL_SESSION_TABLE_SIZE	32
--#define SSL_SESSION_ENTRY_LIFE	86400 * 1000 /* one day */
-+/* #define SSL_SESSION_TABLE_SIZE	32 */
-+/* #define SSL_SESSION_ENTRY_LIFE 	86400 * 1000 */ /* one day */
- 
- /*	Use RFC 5077 session resumption mechanism. The SSL_SESSION_ENTRY_LIFE
- 	define applies to this method as well as the standard method.  The
- 	SSL_SESSION_TICKET_LIST_LEN is the max size of the server key list. */
++
++#ifndef SSL_SESSION_TABLE_SIZE
+ #define SSL_SESSION_TABLE_SIZE 32
+-#define SSL_SESSION_ENTRY_LIFE 86400 * 1000 /* one day */
++#endif
++
++#ifndef SSL_SESSION_ENTRY_LIFE
++#define SSL_SESSION_ENTRY_LIFE         86400 * 1000 /* one */
++#endif
+
+ /*     Use RFC 5077 session resumption mechanism. The SSL_SESSION_ENTRY_LIFE
+        define applies to this method as well as the standard method.  The
+-       SSL_SESSION_TICKET_LIST_LEN is the max size of the server key list. */
++       SSL_SESSION_TICKET_LIST_LEN is the max size of the server key list.
++*/
  #define USE_STATELESS_SESSION_TICKETS
--#define SSL_SESSION_TICKET_LIST_LEN	32
-+/* #define SSL_SESSION_TICKET_LIST_LEN	32 */
- 
++
++#ifndef SSL_SESSION_TICKET_LIST_LEN
+ #define SSL_SESSION_TICKET_LIST_LEN    32
++#endif
+
  /******************************************************************************/
  /*
-@@ -380,6 +380,30 @@ extern "C" {
+@@ -380,6 +390,30 @@ extern "C" {
  */
- /* #define USE_ZLIB_COMPRESSION */	/* @security INSECURE */
- 
+ /* #define USE_ZLIB_COMPRESSION */     /* @security INSECURE */
+
 +/******************************************************************************/
 +/*
-+	Protection against protocol downgrade attacks.
-+	https://tools.ietf.org/html/draft-ietf-tls-downgrade-scsv-00
++       Protection against protocol downgrade attacks.
++       https://tools.ietf.org/html/draft-ietf-tls-downgrade-scsv-00
 +*/
 +
 +/* #define ENABLE_TLS_FALLBACK_SCSV */
 +
 +/******************************************************************************/
 +/*
-+	Partial implementation of the Certificate Status Request extension
-+	https://tools.ietf.org/html/rfc6066#section-8
++       Partial implementation of the Certificate Status Request extension
++       https://tools.ietf.org/html/rfc6066#section-8
 +*/
 +
 +/* #define ENABLE_CERTIFICATE_STATUS_REQUEST */
 +
 +/******************************************************************************/
 +/*
-+	Implementation of the Certificate Transparency extension
-+	(signed_certificate_timestamp)
-+	https://tools.ietf.org/html/rfc6962
++       Implementation of the Certificate Transparency extension
++       (signed_certificate_timestamp)
++       https://tools.ietf.org/html/rfc6962
 +*/
 +
 +/* #define ENABLE_SIGNED_CERTIFICATE_TIMESTAMP */
- 
+
  #ifdef __cplusplus
  }

--- a/inc/patches/matrixssl_matrixsslConfig.h.patch
+++ b/inc/patches/matrixssl_matrixsslConfig.h.patch
@@ -1,5 +1,5 @@
 --- matrixssl-3-7-2b-open-original/matrixssl/matrixsslConfig.h	2015-03-18 20:57:06.000000000 +0000
-+++ matrixssl-3-7-2b-open/matrixssl/matrixsslConfig.h	2016-02-24 21:04:19.000000000 +0000
++++ matrixssl-3-7-2b-open/matrixssl/matrixsslConfig.h	2016-03-30 16:42:19.000000000 +0000
 @@ -61,9 +61,9 @@ extern "C" {
  	the handshake will determine which is best for the connection.
  */
@@ -12,6 +12,24 @@
  #define USE_TLS_RSA_WITH_AES_256_CBC_SHA256 /* TLS 1.2 */
  /* #define USE_TLS_RSA_WITH_AES_128_GCM_SHA256 */ /* TLS 1.2 */
  /* #define USE_TLS_RSA_WITH_AES_256_GCM_SHA384 */ /* TLS 1.2 */
+@@ -224,14 +224,14 @@ extern "C" {
+ 	SSL_SESSION_ENTRY_LIFE is in milliseconds, minimum 0
+ 
+ */
+-#define SSL_SESSION_TABLE_SIZE	32
+-#define SSL_SESSION_ENTRY_LIFE	86400 * 1000 /* one day */
++/* #define SSL_SESSION_TABLE_SIZE	32 */
++/* #define SSL_SESSION_ENTRY_LIFE 	86400 * 1000 */ /* one day */
+ 
+ /*	Use RFC 5077 session resumption mechanism. The SSL_SESSION_ENTRY_LIFE
+ 	define applies to this method as well as the standard method.  The
+ 	SSL_SESSION_TICKET_LIST_LEN is the max size of the server key list. */
+ #define USE_STATELESS_SESSION_TICKETS
+-#define SSL_SESSION_TICKET_LIST_LEN	32
++/* #define SSL_SESSION_TICKET_LIST_LEN	32 */
+ 
+ /******************************************************************************/
+ /*
 @@ -380,6 +380,30 @@ extern "C" {
  */
  /* #define USE_ZLIB_COMPRESSION */	/* @security INSECURE */
@@ -22,7 +40,7 @@
 +	https://tools.ietf.org/html/draft-ietf-tls-downgrade-scsv-00
 +*/
 +
-+#define ENABLE_TLS_FALLBACK_SCSV
++/* #define ENABLE_TLS_FALLBACK_SCSV */
 +
 +/******************************************************************************/
 +/*

--- a/inc/patches/matrixssl_sslDecode.c.patch
+++ b/inc/patches/matrixssl_sslDecode.c.patch
@@ -1,5 +1,5 @@
 --- matrixssl-3-7-2b-open-original/matrixssl/sslDecode.c	2015-03-18 20:57:06.000000000 +0000
-+++ matrixssl-3-7-2b-open/matrixssl/sslDecode.c	2016-03-27 19:43:06.000000000 +0000
++++ matrixssl-3-7-2b-open/matrixssl/sslDecode.c	2016-03-30 16:42:25.000000000 +0000
 @@ -1320,11 +1320,18 @@ static int32 addCompressCount(ssl_t *ssl
  */
  static int32 parseSSLHandshake(ssl_t *ssl, char *inbuf, uint32 len)
@@ -74,7 +74,7 @@
  */
  			if (ssl->myVerifyDataLen == 0) {
  				extLen = 0; /* just overloading existing var */
-@@ -1923,30 +1946,251 @@ SKIP_HSHEADER_PARSE:
+@@ -1923,30 +1946,257 @@ SKIP_HSHEADER_PARSE:
  					if (cipher == TLS_EMPTY_RENEGOTIATION_INFO_SCSV) {
  						ssl->secureRenegotiationFlag = PS_TRUE;
  					}
@@ -176,7 +176,7 @@
 +
 +					if (extType == EXT_SNI) {
 +#ifdef MATRIX_DEBUG
-+						psTraceInfo("Handling SNI extension before selecting cipher");
++						psTraceInfo("Handling SNI extension before selecting cipher\n");
 +#endif
 +						/* Must hold 2 b len 1 b zero 2 b len */
 +						if (extLen < 3 + 2) {
@@ -315,9 +315,15 @@
 +#ifdef MATRIX_DEBUG
 +			if (bestCipherIndex != 32767) {
 +				if (firstCipherIndex != bestCipherIndex) psTraceInfo("Overriding client cipher preferences\n");
-+				psTraceIntInfo("firstCipherIndex = %d\n", firstCipherIndex);
-+				psTraceIntInfo("bestCipherIndex = %d\n", bestCipherIndex);
++				psTraceIntInfo("firstCipherIndex = %d", firstCipherIndex);
++				psTraceIntInfo("  bestCipherIndex = %d", bestCipherIndex);
++				psTraceIntInfo("  bestCipher = %d\n", bestCipher);
++			} else {
++				psTraceInfo("No common cipher found\n");
 +			}
++
++			psTraceIntInfo("Should we use a resumed cipher? %d  ", isResumedCipher);
++			psTraceIntInfo("Resumed cipher: %d\n", resumedCipher);
 +#endif
 +
 +/*
@@ -334,7 +340,7 @@
  /*
  			If we fell to the default cipher suite, we didn't have
  			any in common with the client, or the client is being bad
-@@ -1955,7 +2199,7 @@ SKIP_HSHEADER_PARSE:
+@@ -1955,7 +2205,7 @@ SKIP_HSHEADER_PARSE:
  			if (ssl->cipher == NULL || ssl->cipher->ident != cipher ||
  					cipher == SSL_NULL_WITH_NULL_NULL) {
  				psTraceIntInfo("Can't support requested cipher: %d\n", cipher);
@@ -343,7 +349,7 @@
  				ssl->err = SSL_ALERT_HANDSHAKE_FAILURE;
  				return MATRIXSSL_ERROR;
  			}
-@@ -2051,6 +2295,10 @@ SKIP_HSHEADER_PARSE:
+@@ -2051,6 +2301,10 @@ SKIP_HSHEADER_PARSE:
  						psTraceInfo("Invalid extension header len\n");
  						return MATRIXSSL_ERROR;
  					}
@@ -354,7 +360,24 @@
  #ifdef ENABLE_SECURE_REHANDSHAKES
  /*
  					Handle incoming client extensions we support.
-@@ -2315,6 +2563,14 @@ SKIP_HSHEADER_PARSE:
+@@ -2285,6 +2539,7 @@ SKIP_HSHEADER_PARSE:
+ #ifdef USE_STATELESS_SESSION_TICKETS
+ 					/* Server side parsing */
+ 					if (extType == EXT_SESSION_TICKET) {
++						psTraceInfo("Session ticket extension received\n");
+ 						/* Have a handy place to store this info.  Tickets are
+ 							the	only way a server will make use of 'sid'.
+ 							Could already exist if rehandshake case here */
+@@ -2301,6 +2556,8 @@ SKIP_HSHEADER_PARSE:
+ 							memset(ssl->sid, 0x0, sizeof(sslSessionId_t));
+ 						}
+ 						if (extLen > 0) { /* received a ticket */
++							psTraceInfo("Session ticket received\n");
++
+ 							ssl->sessionIdAndTicket = SESSION_ID_TICKET;
+ 
+ 							if (matrixUnlockSessionTicket(ssl, c, extLen)
+@@ -2315,6 +2572,14 @@ SKIP_HSHEADER_PARSE:
  								matrixsslUpdateStat(ssl, RESUMPTIONS_STAT, 1);
  #endif
  							} else {
@@ -369,7 +392,21 @@
  								/* If client sent a sessionId in the hello,
  									we can ignore that here now */
  								if (ssl->sessionIdLen > 0) {
-@@ -2436,66 +2692,36 @@ SKIP_HSHEADER_PARSE:
+@@ -2335,11 +2600,13 @@ SKIP_HSHEADER_PARSE:
+ #endif
+ 							}
+ 						} else {
++							psTraceInfo("Session ticket request received\n");
+ 							/* Request for session ticket.  Can we honor? */
+ 							if (ssl->keys->sessTickets) {
+ 								ssl->sid->sessionTicketFlag =
+ 									SESS_TICKET_FLAG_RECVD_EXT;
+ 							} else {
++								psTraceInfo("We CAN'T honor it\n");
+ 								ssl->sid->sessionTicketFlag = 0;
+ 							}
+ 						}
+@@ -2436,66 +2703,36 @@ SKIP_HSHEADER_PARSE:
  						}
  						c++; extLen--;
  					}
@@ -458,7 +495,7 @@
  					c += extLen;
  				}
  			}
-@@ -2652,18 +2878,30 @@ SKIP_HSHEADER_PARSE:
+@@ -2652,18 +2889,30 @@ SKIP_HSHEADER_PARSE:
  			need to worry about matching resumed cipher suite.
  */
  			p = c + suiteLen;
@@ -493,7 +530,7 @@
  				ssl->err = SSL_ALERT_HANDSHAKE_FAILURE;
  				psTraceInfo("No matching cipher for SSL handshake\n");
  				return MATRIXSSL_ERROR;
-@@ -3471,7 +3709,7 @@ PROTOCOL_DETERMINED:
+@@ -3471,7 +3720,7 @@ PROTOCOL_DETERMINED:
  				if (memcmp(ssl->sessionId, c, sessionIdLen) == 0) {
  					ssl->flags |= SSL_FLAGS_RESUMED;
  				} else {
@@ -502,7 +539,7 @@
  #ifndef USE_PKCS11_TLS_ALGS
  					memset(ssl->sec.masterSecret, 0x0, SSL_HS_MASTER_SIZE);
  #endif
-@@ -3489,7 +3727,7 @@ PROTOCOL_DETERMINED:
+@@ -3489,7 +3738,7 @@ PROTOCOL_DETERMINED:
  			c += sessionIdLen;
  		} else {
  			if (ssl->sessionIdLen > 0) {
@@ -511,7 +548,7 @@
  #ifndef USE_PKCS11_TLS_ALGS
  				memset(ssl->sec.masterSecret, 0x0, SSL_HS_MASTER_SIZE);
  #endif
-@@ -3524,7 +3762,7 @@ PROTOCOL_DETERMINED:
+@@ -3524,7 +3773,7 @@ PROTOCOL_DETERMINED:
  				return MATRIXSSL_ERROR;
  			}
  		} else {
@@ -520,7 +557,7 @@
  				ssl->err = SSL_ALERT_HANDSHAKE_FAILURE;
  				psTraceIntInfo("Can't support requested cipher: %d\n", cipher);
  				return MATRIXSSL_ERROR;
-@@ -5258,7 +5496,7 @@ static int32 dealWithAlpnExt(ssl_t *ssl,
+@@ -5258,7 +5507,7 @@ static int32 dealWithAlpnExt(ssl_t *ssl,
  	userChoice = MAX_PROTO_EXT;
  
  	/* Already tested if callback exists */
@@ -529,7 +566,7 @@
  
  	if (userChoice == MAX_PROTO_EXT) {
  		/* User chose to completely ignore. No reply extension will be sent */
-@@ -5292,6 +5530,10 @@ static int32 dealWithAlpnExt(ssl_t *ssl,
+@@ -5292,6 +5541,10 @@ static int32 dealWithAlpnExt(ssl_t *ssl,
  	ssl->alpn = proto[userChoice];
  	ssl->alpnLen = protoLen[userChoice];
  
@@ -540,7 +577,7 @@
  	while (i > 0) {
  		if (i - 1 != userChoice) {
  			psFree(proto[i - 1], ssl->sPool);
-@@ -5299,6 +5541,23 @@ static int32 dealWithAlpnExt(ssl_t *ssl,
+@@ -5299,6 +5552,23 @@ static int32 dealWithAlpnExt(ssl_t *ssl,
  		i--;
  	}
  

--- a/lib/Crypt/MatrixSSL3.pm
+++ b/lib/Crypt/MatrixSSL3.pm
@@ -788,6 +788,34 @@ function will use to create the extension data.
 
 =back
 
+Returns the number of files loaded (if this is 0 there was an error loading one of the files).
+
+=head2 refresh_ALPN_data
+
+    $num_protocols = refresh_ALPN_data( $server_index, $index, $protocols );
+
+Used to refresh the application protocols for a default server or for a virtual host.
+
+Parameters:
+
+=over
+
+=item $server_index and $index
+
+Are the same as refresh_OCSP_staple above.
+
+=item $protocols
+
+=over
+
+=item *
+
+Perl array reference containing the new protocols.
+
+=back
+
+Returns the number of protocols you supplied (if this is 0 there was an error loading one of the files).
+
 =back
 
 Returns the number of files loaded in order to build extension data.
@@ -845,6 +873,21 @@ When this object will be destroyed will call:
     matrixSslLoadRsaKeysMem( $keys, $cert, length $cert,
         $priv, length $priv, $trustedCA, length $trustedCA )
 
+=head3 load_ecc
+
+    $rc = $keys->load_ecc( $certFile,
+        $privFile, $privPass, $trustedCAcertFiles );
+
+    matrixSslLoadEcKeys( $keys, $certFile,
+        $privFile, $privPass, $trustedCAcertFiles )
+
+=head3 load_rsa_mem
+
+    $rc = $keys->load_ecc_mem( $cert, $priv, $trustedCA );
+
+    matrixSslLoadEcKeysMem( $keys, $cert, length $cert,
+        $priv, length $priv, $trustedCA, length $trustedCA )
+
 =head3 load_pkcs12
 
     $rc = $keys->load_pkcs12( $p12File, $importPass, $macPass, $flags );
@@ -877,8 +920,8 @@ B<Server side.>
 Return new object $sessID representing (sslSessionId_t*) type.
 Throw exception if failed to allocate memory.
 When this object will be destroyed will free memory, so you should
-keep this object while there are exist Client/Server session
-which uses this $sessID.
+keep this object while there are still Client/Server session
+which use this $sessID.
 
 =head3 clear
 
@@ -1069,7 +1112,7 @@ It has the advantage that the session will contain the latest OCSP data if
 the OCSP DER file is refreshed in the meantime.
 
 Don't be lazy and use $ssl->set_server_params({'OCSP_staple' => '...'}) and
-$ssl->refresh_OCSP_staple() instead.
+$ssl->refresh_OCSP_staple(...) instead.
 
 
 =head2 Crypt::MatrixSSL3::Client and Crypt::MatrixSSL3::Server

--- a/t/07.server_params.t
+++ b/t/07.server_params.t
@@ -10,7 +10,7 @@ Crypt::MatrixSSL3::Open();
 unless (Crypt::MatrixSSL3::capabilities() & OCSP_STAPLES_ENABLED) {
     plan skip_all => "OCSP staples not enabled - OCSP_STAPLES_ENABLED not defined";
 } else {
-    plan tests => 8;
+    plan tests => 9;
 }
 
 my $certFile            = 't/cert/server.crt';
@@ -44,6 +44,7 @@ cmp_ok $Server_SSL->set_server_params($server_index, 420), '==', $server_index, 
 cmp_ok Crypt::MatrixSSL3::refresh_OCSP_staple($server_index, undef, $OCSPtest), '==', PS_SUCCESS, 'Crypt::MatrixSSL3::refresh_OCSP_staple(server_index, undef, file)';
 cmp_ok Crypt::MatrixSSL3::refresh_SCT_buffer($server_index, undef, $CTbuffer), '==', 1, 'Crypt::MatrixSSL3::refresh_SCT_buffer(server_index, undef, file)';
 cmp_ok Crypt::MatrixSSL3::refresh_SCT_buffer($server_index, undef, $CTfiles), '==', 2, 'Crypt::MatrixSSL3::refresh_SCT_buffer(server_index, undef, [files])';
+cmp_ok Crypt::MatrixSSL3::refresh_ALPN_data($server_index, undef, ['proto3', 'proto4']), '==', 2, 'Crypt::MatrixSSL3::refresh_ALPN_data(server_index, undef, [protocols])';
 
 undef $Server_SSL;
 undef $Server_Keys;

